### ISSUE TITLE
Add new role for managing /etc/hosts entries

### ISSUE THREAD

**Issue by rhmk** [sap-linuxlab/community.sap_install#171]
We need correct updates in /etc/hosts several times. So it is valuable to make a separate role out of it.

Example
```
- name: ensure /etc/hosts is updated 
  include_role: sap_update_etc_host
  var:
        sap_update_etc_hosts_name: item.name
        sap_update_etc_hosts_domain: item.domain
        sap_update_etc_host_ip: item.ip
        sap_update_etc_hosts_add_alias: ""
        sap_update_etc_hosts_keep_old_alias: default(true )
        sap_update_etc_hosts_check_dns: default(true)
        sap_update_etc_hosts_dns_error_fatal: false
        sap_update_etc_hosts_name_length: 13 
  loop:
    - { ip: 1.2.3.4, name: node1 , domain: example.com }
    - name: node2
      domain: example.com
      ip: 1.2.3.4
```

Tasks: will be extracted from sap_general_preconfigure

**Comment by berndfinger** [comment]
As per applicable SAP notes (e.g. 2772999), the /etc/hosts entries must be:
ip $(hostname -f) $(hostname -s) [aliases]
So this is something the role must support (just as in sap_general_preconfigure), meaning that we also need to define (or get from ansible) the DNS domain name of the system(s).

I also like the idea of having one variable which contains a list of dicts, but we'll see what Markus comes up with.

**Comment by ja9fuchs** [comment]
We should avoid duplicating definitions. 

Also, in the HSR role I have changed the quite rough solutions I found in other code to a check for conflicts and update only if it's safe. Got experience with wild /etc/hosts in real environments. ;)
https://github.com/sap-linuxlab/community.sap_install/blob/main/roles/sap_ha_install_hana_hsr/tasks/update_etchosts.yml

**Comment by rhmk** [comment]
I agree. The above can then easily be called like:
```
include_role: update_etc_hosts
vars:
   - update_etc_hosts_nodes: "{{ sap_hana_cluster_nodes }}"
```

Janine, Bernd and I are through many possible misconfigurations and the code is basically there. Your two lines are a subset of what is needed. So the idea is to consolidate all these calls with a single role

**Comment by rhmk** [comment]
What final name do you vote for this role:
- update_etc_hosts
- os_update_etc_hosts
- sap_update_etc_hosts 
- anything else?

**Comment by ja9fuchs** [comment]
+1 for a simple "update_etc_hosts"
"os_update...something" could be misleading to be somehow related to OS updates. 
It is not sap special from a functional perspective.

**Comment by kksat** [comment]
I vote for 'etc_hosts' or 'os_etc_hosts'
Because role should be idempotent and not update the records, but make sure correct records are in the /etc/hosts file, also role should remove records from the hosts file sometimes.  For instance like this

```ansible
- name: ensure /etc/hosts is updated 
  include_role: sap_update_etc_hosts
  var:
        sap_update_etc_hosts:
                - ip: 1.2.3.4
                   hostname: host1
                   aliases:
                    - alias1
                    - anotheralias2
                   comment: "Here comes text after hashsign"
                   state: present
                - ip: 1.2.3.5
                   hostname: host2
                   state: absent
```

I've added state parameter compared to my initial suggestion.

**Comment by ja9fuchs** [comment]
All roles should ideally be idempotent, but still are called "configure..." because that is the intention if things are not already correctly in place. :)
I don't think an /etc/hosts update run by our SAP deployment roles should delete anything conflicting from the file. 

Maybe we could make it optional for the user to explicitly accept conflicts to be solved such that new variable values will overwrite existing different entries?

**Comment by sean-freeman** [comment]
Agree with @kksat regarding idempotency and overall "KISS principle" naming convention of Ansible Role (e.g. use `os_` prefix rather than `sap_`).

However I am confused what this Ansible Role does in comparison with:
[community.sap_operations -- /roles/os_etchosts](https://github.com/sap-linuxlab/community.sap_operations/tree/main/roles/os_etchosts)

While update of `/etc/host`s is undoubtably performed during an installation, it is also a maintenance activity. Same as maintaining `~/.ssh/known_hosts` (covered by [community.sap_operations -- /roles/os_knownhosts](https://github.com/sap-linuxlab/community.sap_operations/tree/main/roles/os_knownhosts))

**Comment by rhmk** [comment]
I have named the role os_update_hosts, the input parameters are compatible with the cluster setup parameters and can hence used immediately (see. README) without rewriting

**Comment by sean-freeman** [comment]
@rhmk Can you please address previous comment regarding the correct Ansible Collection for `update_etc_hosts` to belong to? And comparison to existing (incomplete) Ansible Role?
https://github.com/sap-linuxlab/community.sap_install/issues/171#issuecomment-1241756005

**Comment by berndfinger** [comment]
Solved in #495.
